### PR TITLE
Power Plant Costing Updates, v2

### PR DESCRIPTION
**Pull request by bpaul4**
## Changes proposed in this PR:
- Updated currency units handling to call from registered IDAES currency units dictionary
- Added Pyomo unit conversions for currency and basis quantity variables
- Corrected unit name "strings" (e.g. gpm, tpd, MMBtu) to Pyomo syntax so they are parsed correctly (e.g. gal/min, ton/d, MBtu)
- Other minor fixes

### Legal Acknowledgement

By contributing to this software project, I agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the license terms described in the LICENSE.txt file at the top level of this directory.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

**Review by Daison2102 — APPROVED** [review]
These changes look good!

File: idaes/models_extra/power_generation/costing/BB_costing_parameters.json [hunk 1]
```diff
@@ -1 +1 @@
...
..."BEC": 2975.2252770357054, "Eng Fee": 0.15, "Process Contingency": 0.0, "Project Contingency": 0.15, "RP Value": 590.5718256, "Units": "ton/d"}, "10.2": {"BEC": 1685.3960415924944, "Eng Fee": 0.15, "Process Contingency": 0.0, "Project Contingency": 0.15, "RP Value": 590.5718256, "Units": "ton/d"}, "10.3": {"BEC": 757.5010160187344, "Eng Fee": 0.15, "Process Contingency": 0.0, "Project Contingency": 0.15, "RP Value": 590.5718256, "Units": "ton/d"}, "10.6": {"BEC": 2359.4785625848244, "Eng Fee": 0.15, "Process Contingency": 0.0, "Project Contingency": 0.15, "RP Value": 590.5718256, "Units": "ton/d"}, "10.7": {"BEC": 538.7718848531108, "Eng Fee": 0.15, "Process Contingency": 0.0, "Project Contingency": 0.15, "RP Value": 590.5718256, "Units": "ton/d"}, "10.8": {"BEC": 162.2277636863303, "Eng Fee": 0.15, "Process Contingency": 0.0, "Project Contingency": 0.15, "RP Value": 590.5718256, "Units": "ton/d"}, "10.9": {"BEC": 1032.2718711308605, "Eng Fee": 0.15, "Process Contingency": 0.0, "Project Contingency": 0.3, "RP Value": 590.5718256, "Units": "ton/d"}, "11.1": {"BEC": 4773.257359744647, "Eng Fee": 0.15, "Process Contingency": 0.0, "Project Contingency": 0.15, "RP Value": 277700.0, "Uni
```
Comment (MAZamarripa) [inline idaes/models_extra/power_generation/costing/BB_costing_parameters.json]: can you please update the dictionary adding a new key - CE_index_base_year ?
All these accounts are dollars 2018
Maybe this info belongs to BB_costing_exponents.json

File: idaes/models_extra/power_generation/costing/power_plant_capcost.py [hunk 2]
```diff
@@ -678,6 +816,7 @@ def get_PP_costing(
             cost_accounts,
             mutable=True,
             initialize=reference_costs,
+            units=pyunits.USD_2018,
```
Comment (MAZamarripa) [inline idaes/models_extra/power_generation/costing/power_plant_capcost.py]: this should come from the costing dictionary

File: idaes/models_extra/power_generation/costing/power_plant_capcost.py [hunk 3]
```diff
@@ -1051,14 +1219,26 @@ def get_ASU_cost(self, scaled_param):
             fs = self.parent_block()
 
         # build flowsheet level parameters CE_index = year
-        if not hasattr(fs, "costing"):
-            fs.get_costing(year="2017")
+        # if not hasattr(fs, "costing"):
+        #     fs.get_costing(year="2017")
 
-        CE_index = fs.costing.CE_index
+        # CE_index = fs.costing.CE_index
```
Comment (MAZamarripa) [inline idaes/models_extra/power_generation/costing/power_plant_capcost.py]: delete

File: idaes/models_extra/power_generation/costing/power_plant_capcost.py [hunk 4]
```diff
@@ -1381,17 +1602,40 @@ def get_variable_OM_costs(self, production_rate, resources, rates, prices={}):
...
-            "water": 1.90e-3 * pyunits.USD_2018 / pyunits.gallon,
-            "water_treatment_chemicals": 550 * pyunits.USD_2018 / pyunits.ton,
-            "ammonia": 300 * pyunits.USD_2018 / pyunits.ton,
-            "SCR_catalyst": 150 * pyunits.USD_2018 / pyunits.ft**3,
-            "triethylene_glycol": 6.80 * pyunits.USD_2018 / pyunits.gallon,
-            "SCR_catalyst_waste": 2.50 * pyunits.USD_2018 / pyunits.ft**3,
-            "triethylene_glycol_waste": 0.35 * pyunits.USD_2018 / pyunits.gallon,
-            "amine_purification_unit waste": 38 * pyunits.USD_2018 / pyunits.ton,
-            "thermal_reclaimer_unit_waste": 38 * pyunits.USD_2018 / pyunits.ton,
+            "natural_gas": pyunits.convert(4.42 * pyunits.USD_2018, CE_index_base_units)
```
Comment (MAZamarripa) [inline idaes/models_extra/power_generation/costing/power_plant_capcost.py]: not sure you need CE base for natural gas price

File: idaes/models_extra/power_generation/costing/power_plant_capcost.py [hunk 5]
```diff
@@ -1414,32 +1658,32 @@ def get_variable_OM_costs(self, production_rate, resources, rates, prices={}):
 
         # if costing power make vars in fs.costing, if costing hydrogen make vars
         # in fs.H2_costing
-        # if mode == "power":
-        #     costing = self.costing
-        # elif mode == "hydrogen":
-        #     self.H2_costing = Block()
-        #     costing = self.H2_costing
+        if mode == "power":
```
Comment (MAZamarripa) [inline idaes/models_extra/power_generation/costing/power_plant_capcost.py]: why is this back again?

File: idaes/models_extra/power_generation/costing/power_plant_capcost_custom_dict.py [hunk 6]
```diff
@@ -348,7 +348,7 @@
                 "Process Contingency": 0.12,
```
Comment (MAZamarripa) [inline idaes/models_extra/power_generation/costing/power_plant_capcost_custom_dict.py:348]: add CE_base to dictionary exponents

File: idaes/models_extra/power_generation/costing/power_plant_capcost.py [hunk 7]
```diff
@@ -1585,7 +1829,7 @@ def initialize_variable_OM_costs(fs):
             for i in fs.H2_costing.total_variable_OM_cost.keys():
                 calculate_variable_from_constraint(
                     fs.H2_costing.total_variable_OM_cost[i],
-                    fs.total_variable_cost_rule_hydrogen[i],
+                    fs.H2_costing.total_variable_cost_rule_hydrogen[i],
                 )
 
     # -----------------------------------------------------------------------------
```
Comment (MAZamarripa) [inline idaes/models_extra/power_generation/costing/power_plant_capcost.py:1853]: is costing initialization working?

File: idaes/models_extra/power_generation/costing/BB_costing_parameters.json (hunk 1 above)
Comment (bpaul4) [reply idaes/models_extra/power_generation/costing/BB_costing_parameters.json:1]: I'll update BB_costing_parameters.json, the custom_costing_params dictionary and sCO2_costing_params for the appropriate base years. I don't think the exponents need it, they are not dependent on the currency units.

File: idaes/models_extra/power_generation/costing/power_plant_capcost.py (hunk 3 above)
Comment (bpaul4) [reply idaes/models_extra/power_generation/costing/power_plant_capcost.py]: Would you like all of the commented code deleted?

File: idaes/models_extra/power_generation/costing/power_plant_capcost.py (hunk 5 above)
Comment (bpaul4) [reply idaes/models_extra/power_generation/costing/power_plant_capcost.py]: I will revert this back to what it was before, and let you know if any issues come up with the tests.

File: idaes/models_extra/power_generation/costing/tests/test_NGCC_costing.py [hunk 8]
```diff
@@ -234,19 +234,21 @@ def test_units3_costing(build_costing):
...
-    # Obtain circulating water flowrate in gpm
-    cir_water_flowrate = 217555  # gpm
+    # Obtain circulating water flowrate in gal/min
+    cir_water_flowrate = 217555  # gal/min
 
-    m.fs.b12.circ_water_flow = pyo.Var(initialize=cir_water_flowrate)  # gpm
+    m.fs.b12.circ_water_flow = pyo.Var(initialize=cir_water_flowrate)  # gal/min
     m.fs.b12.circ_water_flow.fix()
-    get_PP_costing(m.fs.b12, Circ_water_accounts, m.fs.b12.circ_water_flow, "gpm", 6)
+    get_PP_costing(
```
Comment (MAZamarripa) [inline idaes/models_extra/power_generation/costing/tests/test_NGCC_costing.py]: this import looks like the old costing

File: idaes/models_extra/power_generation/costing/tests/test_NGCC_costing.py [hunk 9]
```diff
@@ -78,10 +78,10 @@ def test_units1_costing(build_costing):
     RW_withdraw_accounts = ["3.2", "3.4", "3.5", "9.5", "14.6"]
     m.fs.b2 = pyo.Block()
 
-    m.fs.b2.raw_water_withdrawal = pyo.Var(initialize=2902)  # gpm
+    m.fs.b2.raw_water_withdrawal = pyo.Var(initialize=2902)  # gal/min
```
Comment (MAZamarripa) [inline idaes/models_extra/power_generation/costing/tests/test_NGCC_costing.py]: need to add pyunits

File: idaes/models_extra/power_generation/costing/tests/test_power_plant_capcost.py [hunk 10]
```diff
@@ -50,7 +52,853 @@ def test_PP_costing():
...
+    # Create a Concrete Model as the top level object
+    m = pyo.ConcreteModel()
+
+    # Add a flowsheet object to the model
+    m.fs = FlowsheetBlock(default={"dynamic": False})
+    m.fs.costing = QGESSCosting()
+
+    # Fixed and Variable Costs:
+    # build variable costs components
+    m.fs.net_power = pyo.Var(m.fs.time, initialize=650, units=pyunits.MW)
```
Comment (MAZamarripa) [inline idaes/models_extra/power_generation/costing/tests/test_power_plant_capcost.py:414]: what happens if we don't pass net_power?

**Comment by bpaul4** [comment]
The grid integration and PySMO test failures are most likely due to a recent `pandas` version change (`numpy` and `matplotlib` are less likely sources of the errors, and also had recent version changes). They are occurring in the main branch as well, so we should not wait on these to merge.

**Comment by bpaul4** [comment]
Closing, since the updates are part of another PR on the main repository.

**Pull request closed without merging**